### PR TITLE
Support vm dynamic scaling with kvm

### PR DESCRIPTION
**Pull request by GutoVeronezi**
### Description
ACS has an API to scale VMs `scaleVirtualMachine`, however dynamic scale of VMs is limited to `XenServer` and `VMWare`.

This PR intends suport live scale for VMs on `KVM`.

When we are using a custom service offerings, KVM's XML only receives the boot memory, but with this changes it will receive the max values defined in the service offerings. If it is a unconstrainced service offering, max values will be the `vm.serviceoffering.cpu.cores.max` and `vm.serviceoffering.ram.size.max` values . If they are `0`, max values will be host or last host capacities.

### Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale
- [ ] Major
- [x] Minor

### How Has This Been Tested?
I created unit tests for the code that is being introduced here. Moreover, it has been tested locally in a test lab. I created multiples VMs with multiples definitions and tested API with they:

| VM STATE | INIT CPU CORES | UP TO | INIT RAM SIZE | UP TO | SERVICE OFFERING | RAM SERVICE OFFERING | CPU SERVICE OFFERING | MAX RAM CONFIG | MAX CPU CONFIG | EXPECTED RESULT |
| -------- | --------- | ----- | -------- | ----- | ---------------- | -------------------- | -------------------- | -------------- | -------------- | --------------- |
| RUNNING | 1 | 3 | 64 | 128 | CONSTRAINED | 256 | 2 | 512 | 4 | FAIL |
| RUNNING | 1 | 2 | 64 | 257 | CONSTRAINED | 256 | 2 | 512 | 4 | FAIL |
| RUNNING | 1 | 2 | 128 | 64 | CONSTRAINED | 256 | 2 | 512 | 4 | FAIL |
| RUNNING | 2 | 1 | 64 | 64 | CONSTRAINED | 256 | 2 | 512 | 4 | FAIL |
| STOPPED | 1 | 3 | 64 | 128 | CONSTRAINED | 256 | 2 | 512 | 4 | FAIL |
| STOPPED | 1 | 2 | 64 | 257 | CONSTRAINED | 256 | 2 | 512 | 4 | FAIL |
| RUNNING | 1 | 2 | 64 | 128 | CONSTRAINED | 256 | 2 | 512 | 4 | SUCCESS |
| RUNNING | 1 | 1 | 64 | 128 | CONSTRAINED | 256 | 2 | 512 | 4 | SUCCESS |
| RUNNING | 1 | 2 | 64 | 64 | CONSTRAINED | 256 | 2 | 512 | 4 | SUCCESS |
| STOPPED | 1 | 2 | 128 | 64 | CONSTRAINED | 256 | 2 | 512 | 4 | SUCCESS |
| STOPPED | 2 | 1 | 64 | 64 | CONSTRAINED | 256 | 2 | 512 | 4 | SUCCESS |
| STOPPED | 1 | 1 | 64 | 128 | CONSTRAINED | 256 | 2 | 512 | 4 | SUCCESS |
| STOPPED | 1 | 1 | 128 | 64 | CONSTRAINED | 256 | 2 | 512 | 4 | SUCCESS |
| RUNNING | 1 | 2 | 128 | 64 | UNCONSTRAINED |  |  | 512 | 4 | FAIL |
| RUNNING | 2 | 1 | 64 | 64 | UNCONSTRAINED |  |  | 512 | 4 | FAIL |
| STOPPED | 1 | 2 | 128 | 64 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| STOPPED | 2 | 1 | 64 | 64 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| RUNNING | 1 | 3 | 64 | 128 | UNCONSTRAINED |  |  | 512 | 4 | FAIL |
| STOPPED | 1 | 1 | 64 | 4000 | UNCONSTRAINED |  |  | 512 | 4 | FAIL |
| RUNNING | 1 | 2 | 64 | 128 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| RUNNING | 1 | 1 | 64 | 128 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| RUNNING | 1 | 2 | 64 | 64 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| RUNNING | 1 | 2 | 64 | 257 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| RUNNING | 1 | 1 | 64 | 700 | UNCONSTRAINED |  |  | 0 | 0 | SUCCESS |
| STOPPED | 1 | 2 | 64 | 128 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| STOPPED | 1 | 1 | 64 | 128 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| STOPPED | 1 | 2 | 64 | 64 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| STOPPED | 1 | 2 | 64 | 257 | UNCONSTRAINED |  |  | 512 | 4 | SUCCESS |
| STOPPED | 1 | 1 | 64 | 700 | UNCONSTRAINED |  |  | 512 | 4 | FAIL |

### CODE REVIEW

**Comment by DaanHoogland** [comment]
¿@wido @weizhouapache @ravening @GabrielBrascher , can you guys look at this please?

**Comment by wido** [comment]
In this PR: https://github.com/apache/cloudstack/pull/4341

The functionality was added that the rootDiskSize can be configured in an Offering.

When we scale the VM we should also scale the rootdisk's size if that is bigger in the new offering.

This PR does not seem to do that. We probably want that added as well.

I really like this PR though. Very much welcome!

**Comment by GutoVeronezi** [comment]
@wido We already have the API `resizeVolume`, which is designed to scale the root disk as well as data disks, and it works with VMs in `stopped` or `running` states.

This API `scaleVirtualMachine` was first designed to scale computing resources (`memory`, `cpuNumber` and `cpuSpeed`). I think we should not mix these two concepts here.

**Comment by wido** [comment]
> @wido We already have the API `resizeVolume`, which is designed to scale the root disk as well as data disks, and it works with VMs in `stopped` or `running` states.
> 
> This API `scaleVirtualMachine` was first designed to scale computing resources (`memory`, `cpuNumber` and `cpuSpeed`). I think we should not mix these two concepts here.

After reading the code and comments from @GabrielBrascher I understand this code only works with custom service offerings, correct?

If so, please disregard my comments.

Otherwise, with PR #4341 merged we now have the option to make the root disk size a condition of the service offering. Thus when changing the offering one needs to make sure that the root disk is also resized according to the new offering.

File: utils/src/main/java/org/apache/cloudstack/utils/bytescale/ByteScaleUtils.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+package org.apache.cloudstack.utils.bytescale;
+
+/**
+ * This class provides a facility to convert bytes through his scales (b, Kib, Kb, Mib, Mb...).
+ *
+ */
+public class ByteScaleUtils {
+
+    public static final int KiB = 1024;
```
Comment (fdalsotto) [inline utils/src/main/java/org/apache/cloudstack/utils/bytescale/ByteScaleUtils.java]: getting myself more familiar with this code base so apologies if the comment isn't good:
could we use org.apache.commons.io.FileUtils instead of hardcoding 1024 in multiple places in the class?

File: server/src/main/java/com/cloud/hypervisor/KVMGuru.java [hunk 2]
```diff
@@ -112,32 +121,147 @@ public VirtualMachineTO implement(VirtualMachineProfile vm) {
...
+    }
+
+    protected Long getVmMaxMemory(ServiceOfferingVO serviceOfferingVO, String vmDescription, Long maxHostMemory) {
+        String serviceOfferingDescription = serviceOfferingVO.toString();
+
+        Long maxMemory;
+        Integer customOfferingMaxMemory = NumberUtils.createInteger(serviceOfferingVO.getDetail(ApiConstants.MAX_MEMORY));
+        Integer maxMemoryConfig = ConfigurationManagerImpl.VM_SERVICE_OFFERING_MAX_RAM_SIZE.value();
+        if (customOfferingMaxMemory != null) {
+            s_logger.debug(String.format("Using 'Custom unconstrained' %s max memory value [%sMb] as %s memory.", serviceOfferingDescription, customOfferingMaxMemory, vmDescription));
```
[the remark this answers is not in the document]
Comment (GutoVeronezi) [reply server/src/main/java/com/cloud/hypervisor/KVMGuru.java]: Thanks for the review.

File: utils/src/main/java/org/apache/cloudstack/utils/bytescale/ByteScaleUtils.java (hunk 1 above)
Comment (GutoVeronezi) [reply utils/src/main/java/org/apache/cloudstack/utils/bytescale/ByteScaleUtils.java]: Actually it is a good point, thanks for the review.

**Comment by GutoVeronezi** [comment]
Can anyone review this?

**Comment by GutoVeronezi** [comment]
Can anyone review this?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 3]
```diff
@@ -4684,11 +4683,15 @@ private VMInstanceVO orchestrateReConfigureVm(String vmUuid, ServiceOffering old
...
+        scaleVmCommand.getVirtualMachine().setType(vm.getType());
 
-        if (vm.getHypervisorType().equals(HypervisorType.VMware)) {
-            HypervisorGuru hvGuru = _hvGuruMgr.getGuru(vm.getHypervisorType());
-            Map<String, String> details = hvGuru.getClusterSettings(vm.getId());
+        final Long dstHostId = vm.getHostId();
+        if(vm.getHypervisorType().equals(HypervisorType.VMware)) {
+            final HypervisorGuru hvGuru = _hvGuruMgr.getGuru(vm.getHypervisorType());
+            Map<String, String> details = null;
+            details = hvGuru.getClusterSettings(vm.getId());
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:4694]: why this change?

**Review by DaanHoogland — APPROVED** [review]
code looks good

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java (hunk 3 above)
Comment (GutoVeronezi) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java]: @DaanHoogland this was a lack of attention. Unnecessary changes undone, thanks.

**Comment by GutoVeronezi** [comment]
@GabrielBrascher could you review this?

**Comment by slavkap** [comment]
@GutoVeronezi, I've tested this with few use cases, and it works as expected. However, I've hit one problem, but not sure if it has to be fixed in this PR:
- Create unconstrained service offering
- disable `enable.dynamic.scale.vm`
- Deploy VM with the new offering
- enable `enable.dynamic.scale.vm`
- update VM with `isdynamicallyscalable=true`
- try to scale the VM 

This error occurs 

> 2021-06-22 07:46:02,730 WARN  [resource.wrapper.LibvirtScaleVmCommandWrapper] (agentRequest-Handler-5:null) (logid:8f9fc53e) Unable to scale com.cloud.agent.api.to.VirtualMachineTO@495dee84 memory to [2097152 KiB] and cpu cores to [4] due to [unsupported configuration: cannot use/hotplug a memory device when domain 'maxMemory' is not defined].
> org.libvirt.LibvirtException: unsupported configuration: cannot use/hotplug a memory device when domain 'maxMemory' is not defined

**Comment by GutoVeronezi** [comment]
@slavkap thanks for testing this feature.

QEMU demands the DOM to be prepared to support CPU scaling and hotplug memory. This preparing is made on VM deploy. ACS verifies if the service offering is dynamic, VM has dynamic scale enabled and global setting `enable.dynamic.scale.vm` as `true`:

https://github.com/apache/cloudstack/blob/607dc234adb16043bd8a2851b56da3e129e90139/server/src/main/java/com/cloud/hypervisor/KVMGuru.java#L180-L186

If they all are enabled and there is a difference between current and max (memory/CPU), the DOM is prepared with the following code:
https://github.com/apache/cloudstack/blob/607dc234adb16043bd8a2851b56da3e129e90139/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVMDef.java#L268-L281

Therefore, if you deploy the VM with the global setting as `false`, it will not prepare the DOM and will result in an error, as `org.libvirt.LibvirtException: unsupported configuration: cannot use/hotplug a memory device when domain 'maxMemory' is not defined`, when trying to scale it.

**Comment by slavkap** [comment]
@GutoVeronezi, I understand your point, but I think this is kind of a problem for the users. I mean that the update of the virtual machine to support dynamic scaling has to be disabled in this case, or when scaling such a VM, the `maxMemory` tag has to be set to the Domain (I don't know if this option is possible)

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapper.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            Domain dm = conn.domainLookupByName(vmName);
+
+            long currentMemory = LibvirtComputingResource.getDomainMemory(dm);
+            long runningVcpus = libvirtComputingResource.countDomainRunningVcpus(dm);
+            long memoryToAttach = newMemory - currentMemory;
+
+            logger.debug(String.format("Scaling %s.", scalingDetails));
+
+            if (memoryToAttach > 0) {
+                String memoryDevice = new LibvirtVmMemoryDeviceDef(newMemory - LibvirtComputingResource.getDomainMemory(dm)).toString();
```
Comment (slavkap) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapper.java:59]:
```suggestion
                String memoryDevice = new LibvirtVmMemoryDeviceDef(memoryToAttach).toString();
```
I think that there is no need to retrieve the domain's memory again

**Comment by GutoVeronezi** [comment]
> @GutoVeronezi, I understand your point, but I think this is kind of a problem for the users. I mean that the update of the virtual machine to support dynamic scaling has to be disabled in this case, or when scaling such a VM, the `maxMemory` tag has to be set to the Domain (I don't know if this option is possible)

As we are changing the settings between deploy and scaling, it creates an inconsistency in the validations.    
About manually adding the `maxMemory` tag at runtime, we would need to change/add some others definitions too, like numa cells in cpu; Also Libvirt do not allow to resize maximum memory on active domains.

Do you have any suggestion about how we can solve this?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapper.java (hunk 4 above)
Comment (GutoVeronezi) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapper.java]: @slavkap done, thanks for the review.

**Comment by slavkap** [comment]
@GutoVeronezi, the only safe option I can think of at the moment is when we update the virtual machine to support dynamic scaling, which is to tell the user that the VM should be stopped (for KVM hypervisors). At the start of the VM, the `maxMemory` will be set.

**Comment by GutoVeronezi** [comment]
@slavkap I added a validation when dynamic scaling the VM and returned a proper message to the user, in case it is not prepared, that I think may direct they.

**Comment by slavkap** [comment]
Thanks, @GutoVeronezi! :)

**Comment by GutoVeronezi** [comment]
@slavkap @DaanHoogland I pushed new changes refactoring `LibvirtScaleVmCommandWrapper` and covering it with better unit tests.

**Comment by slavkap** [comment]
@GutoVeronezi, the code LGTM, but I didn't test it with the latest changes

**Comment by DaanHoogland** [comment]
tests successfull but conflict with new merges. @GutoVeronezi can you have a look?

**Comment by GutoVeronezi** [comment]
@DaanHoogland done, thanks!

**Comment by DaanHoogland** [comment]
@slavkap does this LGTY? cc @nvazquez .

**Comment by slavkap** [comment]
@DaanHoogland, I didn't check the latest changes. I will do it later today and let you know

**Comment by slavkap** [comment]
@DaanHoogland, I've tested with the latest changes and it works as expected. I have only one suggestion, @GutoVeronezi, if you can add in VirtualMachineManagerImpl::reConfigureVM to check for RuntimeException.
```
else if (jobResult instanceof RuntimeException) {
        throw (RuntimeException)jobResult;
}
```

Someone on the mailing list complained that CS does not send proper error msgs to the users in some cases.

> {
>   "accountid": "b3e15775-77f4-11eb-a307-02000a0201ba",
>   "cmd": "org.apache.cloudstack.api.command.admin.vm.ScaleVMCmdByAdmin",
>   "completed": "2021-07-30T10:42:43+0300",
>   "created": "2021-07-30T10:42:42+0300",
>   "jobid": "71c2811b-76b9-47fc-b678-e8b737c46d2e",
>   "jobprocstatus": 0,
>   "jobresult": {
>     "errorcode": 530,
>     "errortext": "Unhandled exception"
>   },
>   "jobresultcode": 530,
>   "jobresulttype": "object",
>   "jobstatus": 2,
>   "userid": "b3e18e06-77f4-11eb-a307-02000a0201ba"
> }

> {
>   "accountid": "b3e15775-77f4-11eb-a307-02000a0201ba",
>   "cmd": "org.apache.cloudstack.api.command.admin.vm.ScaleVMCmdByAdmin",
>   "completed": "2021-07-30T11:10:21+0300",
>   "created": "2021-07-30T11:10:20+0300",
>   "jobid": "b4a42cfb-c37f-43ad-aac1-d07a0800bf06",
>   "jobprocstatus": 0,
>   "jobresult": {
>     "errorcode": 530,
>     "errortext": "Unable to scale vm due to Unable to scale VM {id: \"244\", name: \"i-2-244-VM\", uuid: \"917bea0b-78a2-4923-b0a6-8a056c641cc8\", type: \"User\"} memory to [4194304 KiB] and CPU cores to [4] due to [The VM {id: \"244\", name: \"i-2-244-VM\", uuid: \"917bea0b-78a2-4923-b0a6-8a056c641cc8\", type: \"User\"} is not prepared for dynamic scaling. To be prepared, the VM must be deployed with a dynamic service offering, VM dynamic scale enabled and global setting \"enable.dynamic.scale.vm\" as \"true\". If you changed one of these settings after deploying the VM, consider stopping and starting it again to prepared it to dynamic scaling.]."
>   },
>   "jobresultcode": 530,
>   "jobresulttype": "object",
>   "jobstatus": 2,
>   "userid": "b3e18e06-77f4-11eb-a307-02000a0201ba"
> }

**Comment by GutoVeronezi** [comment]
> @DaanHoogland, I've tested with the latest changes and it works as expected. I have only one suggestion, @GutoVeronezi, if you can add in VirtualMachineManagerImpl::reConfigureVM to check for RuntimeException.
> 
> ```
> else if (jobResult instanceof RuntimeException) {
>         throw (RuntimeException)jobResult;
> }
> ```
> ...

@slavkap Nice catch, done! Thanks.

**Comment by slavkap** [comment]
Thanks, @GutoVeronezi :)

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVmMemoryDeviceDefTest.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2021 The Apache Software Foundation.
```
Comment (GabrielBrascher) [inline plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/LibvirtVmMemoryDeviceDefTest.java:2]: Please check the Copyright

File: plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapperTest.java [hunk 6]
```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2021 The Apache Software Foundation.
```
Comment (GabrielBrascher) [inline plugins/hypervisors/kvm/src/test/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapperTest.java:2]: Please check the **Copyright**

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapper.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2021 The Apache Software Foundation.
```
Comment (GabrielBrascher) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtScaleVmCommandWrapper.java:2]: Please check the **Copyright**

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVmMemoryDeviceDef.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2021 The Apache Software Foundation.
```
Comment (GabrielBrascher) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtVmMemoryDeviceDef.java:2]: @GutoVeronezi I think that these source code headers need to be updated removing  **Copyright** as we discussed in another PR (sorry but I forgot which one xD)

**Review by GabrielBrascher — COMMENTED** [review]
@GutoVeronezi overall it looks good. I've raised a few points to change in regards to the Source code header to meet the Apache license for Apache's projects.

**Comment by GabrielBrascher** [comment]
Manual tests LGTM.

**Comment by GutoVeronezi** [comment]
@GabrielBrascher done, thanks.

**Review by GabrielBrascher — APPROVED** [review]
Thanks for the PR, @GutoVeronezi.
LGTM, based on manual tests and code review.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test
